server/src/main/java/io/crate/action/sql/Cursor.java

```diff
@@ -0,0 +1,131 @@
+    private final List<Symbol> outputs;
+
+    public Cursor(Hold hold,
+                  CompletableFuture<BatchIterator<Row>> queryIterator,
+                  List<Symbol> outputs) {
+        this.hold = hold;
+        this.queryIterator = queryIterator;
+        this.outputs = outputs;
+    }
+
+    public Hold hold() {
+        return hold;
+    }
+
+    public void fetch(RowConsumer consumer, ScrollMode scrollMode, long count) {
+        if (scrollMode == ScrollMode.ABSOLUTE) {
+            consumer.accept(null, new UnsupportedOperationException(
+                "Scrolling to an absolute position is not supported"));
+            return;
+        }
+        if (count < 0) {
+            consumer.accept(null, new UnsupportedOperationException("Cannot scroll backwards"));
+            return;
+        }
+        if (count == 0) {
+            consumer.accept(null, new UnsupportedOperationException("Scroll must move forward"));
+            return;
+        }
+        if (queryIterator.isDone()) {
+            try {
+                BatchIterator<Row> bi = queryIterator.join();
+                triggerConsumer(consumer, bi, count);
+            } catch (Throwable t) {
+                consumer.accept(null, t);
+                return;
+            }
+        } else {
+            queryIterator.whenComplete((bi, err) -> {
+                if (err == null) {
+                    triggerConsumer(consumer, bi, count);
+                } else {
+                    consumer.accept(null, err);
+                }
+            });
+        }
+    }
+
+    private void triggerConsumer(RowConsumer consumer, BatchIterator<Row> fullResult, long count) {
+        // Alternative implementations:
+        // - Could extend RowConsumer to have native scroll support
+        //   e.g. accept(bi, scrollMode, count)?
+        //
+        // - Could also have a "ScrollingBatchIterator"
+
+        BatchIterator<Row> batchIterator;
+        if (count <= Integer.MAX_VALUE) {
+            batchIterator = LimitingBatchIterator.newInstance(fullResult, (int) count);
+        } else {
+            batchIterator = fullResult;
+        }
+        var nonClosingBi = new ForwardingBatchIterator<Row>() {
+
+            @Override
+            protected BatchIterator<Row> delegate() {
+                return batchIterator;
+            }
+
+            @Override
+            public void close() {
+                // Close must be deferred to when the cursor gets closed
+            }
+        };
+        consumer.accept(nonClosingBi, null);
+    }
+
+    public List<Symbol> outputs() {
+        return outputs;
+    }
+
+    @Override
+    public void close() {
+        if (queryIterator.isDone() && !queryIterator.isCompletedExceptionally()) {
+            queryIterator.join().close();
+        } else {
+            queryIterator.whenComplete((bi, err) -> {
+                if (bi != null) {
+                    bi.close();
```

Comment:
clarification - if cursor is declared over a heavy query which terminates mid-way because of RAM counting, its not possible that `queryIterator.isDone() == false &&  bi == null ` --> cursor not being closed, right?

Reply:
If the query terminates mid-way due to RAM accounting the `queryIterator` future would still be completed successfully and the `BatchIterator` would be available.

It only fails if an error happens during the initialization of the query, in which case there is nothing to close either.